### PR TITLE
prototype for callbackOrPromise

### PR DESCRIPTION
**Pull request by amarzavery**
feedback is welcome

### CODE REVIEW

**Comment by amarzavery** [comment]
@balajikris @fiveisprime  - Please review.

File: lib/services/storageManagement2/lib/operations/storageAccounts.js [hunk 1]
```diff
@@ -52,10 +86,34 @@ function StorageAccounts(client) {
...
+        return resolve(result);
+      });
+    });
+  } else {
+    return self._checkNameAvailability(name, options, optionalCallback);
+  }
+};
+
+function _checkNameAvailability(name, options, callback) {
   var client = this.client;
```
Comment (amarzavery) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: I have just prorotyped for checkNameAvailability() method. All the changes from here on in this file are a result of code formatting in vscode. Please ignore them.

Do take a look at the test.js file below.

File: lib/services/storageManagement2/lib/operations/storageAccounts.js [hunk 2]
```diff
@@ -52,10 +86,34 @@ function StorageAccounts(client) {
...
+        
+        return resolve(result);
+      });
+    });
+  } else {
+    return self._checkNameAvailability(name, options, optionalCallback);
+  }
+};
+
+function _checkNameAvailability(name, options, callback) {
```
Comment (amarzavery) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: This is the currently autogenerated method. This will become private.

File: lib/services/storageManagement2/lib/operations/storageAccounts.js [hunk 3]
```diff
@@ -52,10 +86,34 @@ function StorageAccounts(client) {
  *                      {object} [request]  - The HTTP Request object if an error did not occur.
  *
  *                      {stream} [response] - The HTTP Response stream if an error did not occur.
+ * 
+ *          {Promise}   A promise is returned.
+ *             @resolve {CheckNameAvailabilityResult} - The deserialized result object.
+ *             @reject {Error} - The error object.
  */
-StorageAccounts.prototype.checkNameAvailability = function (name, options, callback) {
+StorageAccounts.prototype.checkNameAvailability = function (name, options, optionalCallback) {
```
Comment (amarzavery) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: The method signature remains the same. callback is now optional. If the customer does not provide a callback then we return a Promise.

File: lib/services/storageManagement2/lib/operations/storageAccounts.js [hunk 4]
```diff
@@ -39,10 +41,42 @@ function StorageAccounts(client) {
  * @param {object} [options.customHeaders] Headers that will be added to the
  * request
  *
- * @param {function} callback
+ * @returns {Promise} A promise is returned
+ * @resolve {HttpOperationResponse<CheckNameAvailabilityResult>} - The deserialized result object.
+ * @reject {Error} - The error object.
+ */
+StorageAccounts.prototype.checkNameAvailabilityWithHttpOperationResponse = function (name, options) {
```
Comment (amarzavery) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: The callback used to provide the (err, result, request, response). If someone is interested in it then they can access them.
In Promise land, we can only return one object when resolving the Promise. Hence added this method to return a wrapper object that provides everything.

The standard method below will return a Promise that will resolve with the result (deserialized object).

**Comment by amarzavery** [comment]
@atrauzzi @stevenfollis - Please review. https://github.com/Azure/azure-sdk-for-node/issues/2033

File: lib/services/storageManagement2/lib/operations/storageAccounts.js [hunk 5]
```diff
@@ -39,10 +41,42 @@ function StorageAccounts(client) {
...
+  let self = this;
+  let client = self.client;
+  return new Promise((resolve, reject) => {
+    self._checkNameAvailability(name, options, (err, result, request, response) => {
+      if (err) {
+        return reject(err);
+      }
+      let httpOperationResponse = new HttpOperationResponse(request, response);
+      httpOperationResponse.body = result;
+      return resolve(httpOperationResponse);
```
Comment (atrauzzi) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: May want to confirm this, but I don't think `resolve` and `reject` are required to return anything and are "pure".

Comment (amarzavery) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: That is true.  Have the habit of returning every time from a function while coding. Hence added it. I shall remove it.

Comment (atrauzzi) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: I saw in another project that `return` was being used as a defensive technique inside of promises.  Not sure if that changes things here or not.

Comment (amarzavery) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: Well I tested the code by removing the return keyword and added a `console.log()`. The result was amazing: `console.log()` got executed.

### Modified method by removing the "return" keyword 
```javascript
StorageAccounts.prototype.checkNameAvailability = function (name, options, optionalCallback) {
  var client = this.client;
  var self = this;
  if (!optionalCallback && typeof options === 'function') {
    optionalCallback = options;
    options = null;
  }
  if (!optionalCallback) {
    return new Promise((resolve, reject) => {
      self._checkNameAvailability(name, options, (err, result, request, response) => {
        if (err) reject(err);
        console.log('Before resolving...');
        console.log(result);
        resolve(result);  //<<<<<<
        console.log('After resolving...');
      });
    });
  } else {
    return self._checkNameAvailability(name, options, optionalCallback);
  }
};
```

### The test code that called the method
```javascript
msRestAzure.loginWithServicePrincipalSecret(spn, secret, domain, function (err, credentials) {
  var client = new storageManagementClient(credentials, subscriptionId);
  let name = 'foo'; 
  //calling the method in promise style
  client.storageAccounts.checkNameAvailability(name).then(function (result) {
    console.log('In the thenable waiting for the result of the resolved promise..');
    console.log(result);
  }).catch(function (err) {
    console.log(err);
  });
});
```

##### Output:
```bash
bash-3.2$ node test.js
Before resolving...
{ nameAvailable: false,
  reason: 'AlreadyExists',
  message: 'The storage account named foo is already taken.' }
After resolving... //<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
In the thenable waiting for the result of the resolved promise..
{ nameAvailable: false,
  reason: 'AlreadyExists',
  message: 'The storage account named foo is already taken.' }
```

##### Output after adding the "return" keyword back to the resolve statement:

```bash
bash-3.2$ node test.js
Before resolving...
{ nameAvailable: false,
  reason: 'AlreadyExists',
  message: 'The storage account named foo is already taken.' }
In the thenable waiting for the result of the resolved promise..
{ nameAvailable: false,
  reason: 'AlreadyExists',
  message: 'The storage account named foo is already taken.' }
bash-3.2$
```

### TLDR; return is required based on the above test.

What do you think?

Comment (amarzavery) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: Node version on the system is: 6.9.5

Comment (atrauzzi) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: That makes sense because the callback doesn't end the current function.  So yeah, it might be good evening just based on the semantics of your expectation for execution to end to keep the return statements.

Comment (stepro) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: The right semantics here is really to call reject or resolve as one statement and then ensure the function ends after that (whether through an explicit return or it just happens to leave the function).

```
if (err) {
    reject(err);
} else {
    resolve(result);
}
```

Or even simpler:

```
err ? reject(err) : resolve(result)
```

**Comment by stepro** [comment]
Changes look good to me.

File: lib/services/storageManagement2/lib/operations/storageAccounts.js (hunk 5 above)
Comment (atrauzzi) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: Agreed!

Comment (amarzavery) [inline lib/services/storageManagement2/lib/operations/storageAccounts.js]: I see. Shall follow that.

**Comment by amarzavery** [comment]
@mattchr @xingwu1 - Please review this prototype to support callbacks and promises. Please include Pablo and other team mates with JS expertise on this review. We are updating the generator to make these changes and plan to roll out these changes soon.

**Comment by amarzavery** [comment]
@atrauzzi  A new version of the SDK has been published. Please check the [release page](https://github.com/Azure/azure-sdk-for-node/releases/tag/2.0.0-preview-April2017).
